### PR TITLE
[IMPROVE] Added auto-focus for better user-experience.

### LINKED ISSUE #19378: [Improve] Focus should be in search fields or in input text box of modal windows
### Description:

After user go to Directory now focus is not on search field and one more move is necessary. 
In general - if any new window appeared with search field or new modal window with input text box - focus should be there

### Steps to reproduce:

1. Go to Directory
2. Start type search query
3. Nothing happens because focus is not in search field

### Expected behavior:

Focus should be in a search filed right after user opens Directory

### Actual behavior:

No text cursor
image

### Server Setup Information:

- Version of Rocket.Chat Server: 3.7.1
- Operating System: CentOS7
- Deployment Method: docker
- Number of Running Instances: 12
- DB Replicaset Oplog: Enabled
- NodeJS Version: v12.18.4
- MongoDB Version: 4.0.19

### Client Setup Information

- Desktop App or Browser Version: 3.0.4
- Operating System: Windows 10

### Additional context

Issue could be here
https://github.com/RocketChat/Rocket.Chat/blob/833cb1ef1eaf3b69108de494a5e61f71cf03ccb2/app/ui/client/views/app/components/Directory/DirectoryTable.js#L82-L84

### Relevant logs:

No

### PR DESCRIPTION
**Pull request by Darshilp326**
## Issue(s)

Closes #19378 
While sending file message there is no focus on upload file modal.
While opening contexual-bar options like threads,discussions,members no focus in search bars.
## Steps to test or reproduce

https://user-images.githubusercontent.com/55157259/103024875-af091e80-4576-11eb-8ad0-04e6b4954fbb.mp4

https://user-images.githubusercontent.com/55157259/103025023-fee7e580-4576-11eb-9e7e-3348691200d1.mp4

### CODE REVIEW

**Comment by Darshilp326** [comment]
Can anyone help me to solve failed check?

**Comment by Darshilp326** [comment]
Please review! :)

File: client/components/FilterByText.tsx [hunk 1]
```diff
@@ -36,7 +36,7 @@ const FilterByText: FC<FilterByTextProps> = ({
 	}, []);
 
 	return <Box mb='x16' is='form' onSubmit={handleFormSubmit} display='flex' flexDirection='row' {...props}>
-		<TextInput placeholder={placeholder ?? t('Search')} addon={<Icon name='magnifier' size='x20'/>} onChange={handleInputChange} value={text} />
+		<TextInput placeholder={placeholder ?? t('Search')} autoFocus={true} addon={<Icon name='magnifier' size='x20'/>} onChange={handleInputChange} value={text} />
```
Comment (RonLek) [inline client/components/FilterByText.tsx:39]: Just adding `autoFocus` here would do the trick.

File: client/views/room/contextualBar/Discussions/index.js [hunk 2]
```diff
@@ -213,7 +213,7 @@ export function DiscussionList({ total = 10, discussions = [], loadMoreItems, lo
 		</VerticalBar.Header>
 		<VerticalBar.Content paddingInline={0}>
 			<Box display='flex' flexDirection='row' p='x24' borderBlockEndWidth='x2' borderBlockEndStyle='solid' borderBlockEndColor='neutral-200' flexShrink={0}>
-				<TextInput placeholder={t('Search_Messages')} value={text} onChange={setText} addon={<Icon name='magnifier' size='x20'/>}/>
+				<TextInput placeholder={t('Search_Messages')} value={text} autoFocus={true} onChange={setText} addon={<Icon name='magnifier' size='x20'/>}/>
```
Comment (RonLek) [inline client/views/room/contextualBar/Discussions/index.js:216]: Just adding `autoFocus` here would do the trick.

File: client/views/room/contextualBar/RoomMembers/List/RoomMembers.js [hunk 3]
```diff
@@ -102,7 +102,7 @@ export const RoomMembers = ({
 				<Box display='flex' flexDirection='row' p='x12' flexShrink={0}>
 					<Box display='flex' flexDirection='row' flexGrow={1} mi='neg-x4'>
 						<Margins inline='x4'>
-							<TextInput placeholder={t('Search_by_username')} value={text} onChange={setText} addon={<Icon name='magnifier' size='x20'/>}/>
+							<TextInput placeholder={t('Search_by_username')} autoFocus={true} value={text} onChange={setText} addon={<Icon name='magnifier' size='x20'/>}/>
```
Comment (RonLek) [inline client/views/room/contextualBar/RoomMembers/List/RoomMembers.js:105]: Just adding `autoFocus` here would do the trick.

File: client/views/room/contextualBar/Threads/index.js [hunk 4]
```diff
@@ -253,7 +253,7 @@ export function ThreadList({ total = 10, threads = [], room, unread = [], unread
 			<Box display='flex' flexDirection='row' p='x24' borderBlockEndWidth='x2' borderBlockEndStyle='solid' borderBlockEndColor='neutral-200' flexShrink={0}>
 				<Box display='flex' flexDirection='row' flexGrow={1} mi='neg-x4'>
 					<Margins inline='x4'>
-						<TextInput placeholder={t('Search_Messages')} value={text} onChange={setText} addon={<Icon name='magnifier' size='x20'/>}/>
+						<TextInput placeholder={t('Search_Messages')} autoFocus={true} value={text} onChange={setText} addon={<Icon name='magnifier' size='x20'/>}/>
```
Comment (RonLek) [inline client/views/room/contextualBar/Threads/index.js:256]: Just adding `autoFocus` here would do the trick.

File: app/ui/client/lib/fileUpload.js [hunk 5]
```diff
@@ -260,7 +260,7 @@ export const fileUpload = async (files, input, { rid, tmid }) => {
 			confirmButtonText: t('Send'),
 			cancelButtonText: t('Cancel'),
 			html: true,
-			onRendered: () => $('#file-name').focus(),
+			onRendered: () => $('#file-description').focus(),
```
Comment (RonLek) [inline app/ui/client/lib/fileUpload.js:263]: I think this should not be changed, since the user might want to edit the name of the file before upload.

**Comment by RonLek** [comment]
@Darshilp326 left few comments :)

File: app/ui/client/lib/fileUpload.js (hunk 5 above)
Comment (Darshilp326) [reply app/ui/client/lib/fileUpload.js:263]: The mentioned issue told to focus on description.

File: app/ui-utils/client/lib/modal.js [hunk 6]
```diff
@@ -209,16 +209,14 @@ Template.rc_modal.helpers({
...
-
 	if (this.data.input) {
 		$('.js-modal-input', this.firstNode).focus();
 	} else if (this.data.showConfirmButton && this.data.confirmOnEnter) {
 		$('.js-confirm', this.firstNode).focus();
 	}
-
+	if (this.data.onRendered) {
+		this.data.onRendered();
+	}
```
[the remark this answers is not in the document]
Comment (Darshilp326) [reply app/ui-utils/client/lib/modal.js:219]: Actually above if else if part takes away autofocus from onRendered function.

**Comment by Darshilp326** [comment]
@MartinSchoeler Please review :)

**Review by MartinSchoeler — CHANGES_REQUESTED** [review]
Hey @Darshilp326 unfortunately this pull request is still not ideal, firstly because this will hardcode the autofocus on components that are used in other parts of Rocket.Chat, and this can lead to unwanted behaviors. Second, the native autofocus prop has it's limitations, so my suggestion would be to use our `useAutoFocus` hook for that.

I've created a pull request fixing only the directory screen, it should serve as a good example/base for the other elements that you are trying to add autofocus.

Here is the Pull Request https://github.com/RocketChat/Rocket.Chat/pull/20509

Feel free to contact me if you have any questions about my implementation. 

Looking forward to see the new changes. Thanks!

**Comment by Darshilp326** [comment]
@MartinSchoeler Okay got your point! I will be making new changes to this pr:)

**Review by MartinSchoeler — APPROVED** [review]
Looks good to me now!

**Pull request merged**